infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/RuntimeConsistencyChecker.java

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2017 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.workspace.infrastructure.docker;
+
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.config.Environment;
+import org.eclipse.che.api.core.model.workspace.config.MachineConfig;
+import org.eclipse.che.api.core.model.workspace.runtime.Machine;
+
+import java.util.Map;
+
+/** Checks whether runtime is consistent with its configuration. */
+class RuntimeConsistencyChecker {
+    void check(Environment environment, DockerInternalRuntime runtime) throws ValidationException {
+        Map<String, ? extends MachineConfig> configs = environment.getMachines();
+        Map<String, ? extends Machine> machines = runtime.getMachines();
+        if (configs.size() != machines.size()) {
```

Comment:
I mean that not every machine from env recipe has a matching entity in machines field of environment. But normalization may fix it. I'm not sure here.